### ISSUE TITLE
android not work

### ISSUE THREAD

**Issue by lovekill** [srikanth-lingala/zip4j#148]
Cannot work on android that does not support file.topath (). But it's okay before the update

**Comment by srikanth-lingala** [comment]
which version of Android have to tested on?

**Comment by srikanth-lingala** [comment]
After a bit of research, I think you are using Android v 7.1 or less. [`File.toPath()`][1] is supported from API level 26 (Android v 8.0). The new zip4j update uses a lot of `File.toPath()` for some newly introduced features. I will analyse a work around for it and try to come up with a fix. But this will most likely mean losing those features in Android version <= 7.1, as without that api, it will not be possible to support those features, for example: support for Symlinks. 

[1]: https://developer.android.com/reference/java/io/File#toPath()

**Comment by nicolabeghin** [comment]
@srikanth-lingala if it can help, here's basic android project to test `zip4j` (it's the one I used in the past)

https://github.com/nicolabeghin/zip4jtest

**Comment by srikanth-lingala** [comment]
I have made a fix for this issue, but I have made only a snapshot release for now. Can you please test with version `2.4.1-SNAPSHOT` and let me know if it works fine now? If you have any trouble to get this snapshot version from maven, please have a look at [this answer][1].

[1]: https://stackoverflow.com/a/7717234/2736153

**Comment by nicolabeghin** [comment]
@lovekill are you interested anymore? If not, I'll pick it up since we're using log4j on Android.

**Comment by srikanth-lingala** [comment]
@nicolabeghin Yes, I will reopen. It would be great if you can test and let me know if all works fine (especially including the new feature to rename a file in zip).

**Comment by nicolabeghin** [comment]
I think the best long-term solution will be to run all the tests under Android. Could you kindly upload the 2.4.1 sources on another branch so that I can check it out and set up an Android project comprehensive of unit testing?

**Comment by nicolabeghin** [comment]
Updated repo https://github.com/nicolabeghin/zip4j-android-test with ready-to-use structure able to pull latest zip4j lib and run tests against Android runtime. Here's test results against stable version `2.4.0`: 

[Test Results - zip4j_in_zip4j.html.zip](https://github.com/srikanth-lingala/zip4j/files/4289893/Test.Results.-.zip4j_in_zip4j.html.zip)
![image](https://user-images.githubusercontent.com/2743637/75927482-bdae0a00-5e6c-11ea-9a0e-4a5de06e6f85.png)

**Comment by lovekill** [comment]
@srikanth-lingala @nicolabeghin  
I have other problem 。when I  add fille in zip ,will get NullPointException. but not All zip . 
reappear it  steps:

1. download this file  [http://oss.youxifan.com/MGInc6E](url) .yes it's  apk .(apk is a zip)
2. use the add Api code .
`ZipFile zipFile = new ZipFile("/Users/engine/apk/test.apk");
    ArrayList<File> addFiles = new ArrayList<>();
    addFiles.add(new File("/Users/engine/apk/test.txt"));
    ZipParameters parameters = new ZipParameters();
    parameters.setCompressionMethod(CompressionMethod.DEFLATE);
    parameters.setCompressionLevel(CompressionLevel.NORMAL);
    // 目标路径
    parameters.setRootFolderNameInZip("META-INF/");
    zipFile.addFiles(addFiles, parameters);`
3.  you will get  java.lang.NullPointerException and  damage the apk
`Stack trace:
 net.lingala.zip4j.exception.ZipException: java.lang.NullPointerException
	at net.lingala.zip4j.headers.HeaderWriter.writeFileHeader(HeaderWriter.java:501)
	at net.lingala.zip4j.headers.HeaderWriter.writeCentralDirectory(HeaderWriter.java:382)
	at net.lingala.zip4j.headers.HeaderWriter.finalizeZipFile(HeaderWriter.java:177)
Caused by: java.lang.NullPointerException
	at java.io.OutputStream.write(OutputStream.java:75)
	at net.lingala.zip4j.headers.HeaderWriter.writeRemainingExtraDataRecordsIfPresent(HeaderWriter.java:543)
	at net.lingala.zip4j.headers.HeaderWriter.writeFileHeader(HeaderWriter.java:495)
	... 37 more
`

4.  I have fix this Exception  by modify  HeaderWriter.writeRemainingExtraDataRecordsIfPresent() .at the latest code :
`if(extraDataRecord.getData()!=null){
        outputStream.write(extraDataRecord.getData());
      }`

**Comment by srikanth-lingala** [comment]
@nicolabeghin Thanks for setting up tests for Android. Appreciate it. I will try to get it integrated into the build lane on travis.

**Comment by srikanth-lingala** [comment]
@lovekill I will have a look at the issue.

**Comment by srikanth-lingala** [comment]
@lovekill The zip file is 1.1 GB. Is it for real that an apk is of 1.1 GB in size or do I have a wrong link?

**Comment by srikanth-lingala** [comment]
@lovekill I have create a new issue for this to make tracking easier. Please follow issue #154  for updates

**Comment by nicolabeghin** [comment]
@srikanth-lingala not sure this can be automated easily, but it should be possible! In the meanwhile, if you could share the aforementioned branch I could help out with the tests (manually running them in Android Studio)

**Comment by srikanth-lingala** [comment]
@nicolabeghin If I understand you right, in the new branch you want to do something like in the [zip4j-android-test][1] project, and make it an Android project so that I can run them in Android studio, right? If true, the questions I have is that I probably have to sync master with this branch to make sure that code base is up-to-date, and also I have to be running them manually on Android studio, right?

The way I am currently imagining a solution for this is to have a build run the same tests against JDK and also Android SDK which are in master.  Although honestly speaking, in the few minutes I spent researching this, it does not look encouraging, and I am not quite sure if this will be possible to do. 

As you are an Android developer, maybe you have some tips/suggestions if a) it is even possible to run the tests in the way of my proposal (run tests against Android SDK on travis) and b) how are Android builds tested in general. We do not want an Android app testing, but just to use Android SDK to test against instead of JDK. Would that even be possible? I hope so because then I do not have to manually run the tests after commits (and also I don't have to install Android studio on my machine).

BTW, I fixed those NPE in the tests from your screenshot.

[1]: https://github.com/nicolabeghin/zip4j-android-test

**Comment by nicolabeghin** [comment]
Hi @srikanth-lingala I don't want to touch the code, just perform testing against it (since I don't have access to the SNAPSHOT-2.4.1 branch). After we get this done, I'll try to understand how to automatize the process (always in another/related project, since I don't see a way to "unify" this - I'm thinking to a docker image BTW).

**Comment by srikanth-lingala** [comment]
@nicolabeghin The fix which is part of SNAPSHOT 2.4.1 is in the master branch. You can run the tests against master branch. [Here][1] is the commit from the master with the fix.

[1]: https://github.com/srikanth-lingala/zip4j/commit/c2ea650148ea7a1a5a8cee7f6476332c572c741e

**Comment by nicolabeghin** [comment]
@srikanth-lingala updated repo https://github.com/nicolabeghin/zip4j-android-test/ with CircleCI Continuous Integration based on Docker Android images - don't know if you can embed it somehow in the standard CI flow. Otherwise, you can just embed the building badget from the repo in order to have at-a-glance build status on Android. It pulls from repo, build and run tests against Android (API 27). 

I'll check how to add more build on older Android levels.

**Comment by srikanth-lingala** [comment]
I want to ignore one long running test class for Android. Is there a way that I can ignore all tests in the class `ZipFileZip64IT`? Usually in gradle you can exclude tests by declaring them in `build.gradle`, but unfortunately that does not seem to work in this Android configuration.

**Comment by nicolabeghin** [comment]
Yes you can decide based on namespace or class AFAIK. [Here](https://github.com/nicolabeghin/zip4j-android-test/blob/master/.circleci/config.yml#L29)'s where I declare to run test against all the JUnit classes of package `net.lingala.zip4j.*` but it's possible to define specific classes only.

**Comment by srikanth-lingala** [comment]
I tried a lot. With Java+gradle it is straight forward. But with Android flavour, it is getting quite complicated. I could have made the tests quite faster by excluding this one test class from Android tests (they will be tested under the regular build). But I have already spent too much time on it, and I have a few more things to take care of. If you by any chance get it working, I would appreciate it. Or else, I will have a look at it later.

**Comment by srikanth-lingala** [comment]
so I managed to setup a [CircleCI workflow build][1] (similar to what you have). This workflow will be triggered by the current travis build before it starts, so that both of them will run in parallel. It would still be good to exclude that one Test class in Android tests. 

The bug mentioned in this issue is resolved, and I will make a release soon.

[1]: https://app.circleci.com/pipelines/github/srikanth-lingala/zip4j-android-test?branch=master

**Comment by srikanth-lingala** [comment]
Fix included in v2.5.0 released today. A couple of more features have been added in this release. Please have a look at [release notes][1].

[1]: https://github.com/srikanth-lingala/zip4j/releases

**Comment by nicolabeghin** [comment]
Don't know if you saw this already, but with latest release 2.5.0 an higher number of Android APIs are failing tests:

![image](https://user-images.githubusercontent.com/2743637/76168535-10d7d380-6171-11ea-8424-2f4b1fdef382.png)

**Comment by srikanth-lingala** [comment]
I don't know why they are failing now, but they were all green during the deployment: https://app.circleci.com/pipelines/github/srikanth-lingala/zip4j-android-test/21/workflows/24b0efe8-07cb-426b-9c7b-8752542b4e36

In general, the tests on CircleCI have been flaky earlier because of one flaky test. But this time, it seems to be that the tests are taking too long for some reason and the tests are timing out after 10 mins of no output. I am pretty certain that there was no code change which broke tests (tests were green earlier during the deployment), but I guess some random issue either on Circle CI side or something else. I will check this again tomorrow.

But the release v2.5.0 should be fine.

**Comment by nicolabeghin** [comment]
I confirm [CircleCI has a 10m default timeout](https://support.circleci.com/hc/en-us/articles/360007188574-Build-has-hit-timeout-limit). I just changed the timeout to 30m in https://github.com/nicolabeghin/zip4j-android-test/commit/cfa35cae1b53379fabb64c7500eb706b3f5221af will report back if all good!

**Comment by nicolabeghin** [comment]
As a closing note: I confirm excluding that specific test method all tests pass on the different Android levels! feel free to pick up my little change in the update script, so that Android tests can be performed automatically for each new zip4j release! also it would be extremely useful to have this badge in the repo main page.

**Comment by srikanth-lingala** [comment]
I have synced your changes to my fork and excluded the random failing test. Along with that, I also excluded Zip64 test for Android. This test will be run in the regular build in Travis. This is a very long running task (and hence long tests runtime), and has nothing specific to Android. With this exclusion, we do not need 30 minutes wait time for tests anymore on CircleCI. 

I have added a badge on zip4j's repo readme to indicate also the status of Android tests. 

Thanks again for helping me with this @nicolabeghin.  Appreciate it.
